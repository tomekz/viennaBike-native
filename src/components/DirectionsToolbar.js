import React, { Component } from 'react';
import GeolocationProvider from '.././lib/GeolocationProvider'
import { View, Text, TouchableOpacity, Linking } from 'react-native'
import Icon  from 'react-native-vector-icons/MaterialCommunityIcons';
import styles from './styles/DirectionsToolbar'
import { cWhite }  from '../styles/styles'

class DirectionsToolbar extends Component {

  constructor(props){
    super()
    this.state = {
      show: false,
      source: null,
      destination: null
    }
  }

  async show(destination){
    try{
      const position = await new GeolocationProvider().getPosition()
      const source = { latitude: position.coords.latitude, longitude: position.coords.longitude}
      this.setState({show: true, source: source, destination: destination})
    }
    catch(err){
      this.setState({show: false})
    }
  }

  hide(){
    this.setState({show: false})
  }

  getParameterString = (params) => {
    return params
      .map(({ key, value }) => {
        const encodedKey = encodeURIComponent(key)
        const encodedValue = encodeURIComponent(value)

        return `${encodedKey}=${encodedValue}`
      })
      .join('&')
  }

  getDirections = () => {
    const params = []
    const { source, destination } = this.state

    params.push({
      key: 'saddr',
      value: `${source.latitude},${source.longitude}`
    })

    params.push({
      key: 'daddr',
      value: `${destination.latitude},${destination.longitude}`
    })

    const url = `http://maps.google.com/maps?${this.getParameterString(params)}`

    return Linking.canOpenURL(url).then(supported => {
      if (!supported) {
        return Promise.reject(new Error(`Could not open the url: ${url}`))
      } else {
        return Linking.openURL(url)
      }
    })
  }

  render(){
    if (this.state.show){
      return(
          <TouchableOpacity
          style={styles.container}
          onPress={this.getDirections.bind(this)}
          >
            <Text style = {styles.text}>
              Get directions
            </Text>
            <Icon
              name="directions"
              size={30}
              color={cWhite}
            />
          </TouchableOpacity>
      )
    }
    return null;
  }
}

export default DirectionsToolbar;
