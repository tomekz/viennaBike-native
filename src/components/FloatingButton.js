import React, { Component } from 'react';
import { Text, View, TouchableOpacity } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import styles  from './styles/FloatingButton'

class FloatingButton extends Component {

  onPressButton(){
    this.props.navigation.navigate('DrawerOpen')
  }

  render() {
    return (
        <TouchableOpacity onPress={this.onPressButton.bind(this)} style={styles.button}>
          <Ionicons
            name="md-menu"
            size={24}
          />
        </TouchableOpacity>
    );
  }
}

export default FloatingButton;
