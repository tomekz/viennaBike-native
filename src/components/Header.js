import React, { Component } from 'react';
import { Text, View, TouchableOpacity } from 'react-native';
import { NavigationActions } from 'react-navigation'
import { Ionicons } from '@expo/vector-icons';
import styles  from './styles/Header'
import { colorWhite }  from '../styles/styles'

class Header extends Component {


  onPressMenuButton(){
    this.props.navigation.navigate('DrawerOpen')
  }

  render() {
    return (
      <View style={styles.headerContainer}>
        <TouchableOpacity
          style={styles.headerIcon}
          onPress={this.onPressMenuButton.bind(this)}
        >
          <Ionicons
            name="md-menu"
            size={30}
            color={colorWhite}
          />
        </TouchableOpacity>
        <Text style = {styles.headerTitle}>
          CityBike Vienna
        </Text>

        {this.props.showRefreshButton ?
        <TouchableOpacity
         style={styles.headerIcon}
         onPress={this.props.onRefreshPress}
        >
          <Ionicons
            name="md-refresh"
            size={30}
            color={colorWhite}
          />
        </TouchableOpacity>
          : null }
      </View>
    );
  }
}

export default Header;
