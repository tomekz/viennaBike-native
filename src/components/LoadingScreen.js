import React from 'react';
import { ActivityIndicator, View } from 'react-native';
import styles from './styles/LoadingScreen';

const LoadingScreen = ({ color }) => (
  <View style={styles.root}>
    <ActivityIndicator
      size="large"
      color={color || 'gray'}
    />
  </View>
);

export default LoadingScreen;
