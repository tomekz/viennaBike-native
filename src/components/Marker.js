import React, { Component } from 'react';
import { MapView } from 'expo';

class Marker extends React.Component {
  constructor(){
    super()
    this.marker = null
  }

  render () {
    return (
      <MapView.Marker {...this.props} ref={ref => {this.marker = ref}} />
    )
  }

  componentDidMount() {
    this.updateCallout()
  }

  updateCallout () {
    if (this.props.calloutVisible) {
      this.marker.showCallout()
    } else {
      this.marker.hideCallout()
    }
  }
}

export default Marker
