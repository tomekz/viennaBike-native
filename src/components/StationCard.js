import React, { Component } from 'react';
import { View, ListView, Text, TouchableOpacity } from 'react-native'
import { Avatar} from 'react-native-elements'
import styles from './styles/StationsCard'
import commonStyles from '.././styles/styles'

class StationCard extends Component {

  onPressCard() {
    this.props.onPressCard(this.props.station)
  }

  render()
  {
    const { station } = this.props
    return (
        <TouchableOpacity
        onPress={this.onPressCard.bind(this)}
        style={styles.card}>
          <View style={styles.cardTop}>
            <Avatar
              containerStyle = {styles.cardAvatar}
              small
              rounded
              source={require('../assets/img/citybike_logo.png')}
              overlayContainerStyle={{backgroundColor: commonStyles.colorWhite}}
            />
            <View style={styles.cardTopContent}>
              <Text style={commonStyles.textRegular}>
                Station {station.extra.internal_id}
              </Text>
              <Text style={commonStyles.textLightGrey}>
                {station.name}
              </Text>
            </View>
          </View>
          <View style={styles.cardBottom} >
              <Text style={[commonStyles.textRegular, station.free_bikes == 0 && commonStyles.textRed]} >
               {station.free_bikes} bikes
              </Text>
              <Text style={commonStyles.textRegular}  >
               {station.extra.slots} slots
              </Text>
               { station.distance &&
                <Text style={commonStyles.textBold} >
                  {station.distance} km
                </Text>
              }
          </View>
        </TouchableOpacity>
    )
  }
}

export default StationCard;
