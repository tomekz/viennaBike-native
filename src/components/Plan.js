import React, { Component } from 'react';
import { View, Text } from 'react-native';
import MapView from 'react-native-maps';
import { StationCard, DirectionsToolbar }  from '../components';
import styles from './styles/Plan'

const LATITUDE_DELTA = 0.00922*1.5
const LONGITUDE_DELTA =  0.00421*1.5
const DEFAULT_SELECTED_STATION_ID = '111'  // Oper in center

class Plan extends Component {

  constructor(props){
    super()
    const selectedStation = props.selectedStation || props.stations.find( s=> s.extra.uid === DEFAULT_SELECTED_STATION_ID)
    const region = {
      latitude:       selectedStation.latitude,
      longitude:      selectedStation.longitude,
      latitudeDelta:  LATITUDE_DELTA,
      longitudeDelta: LONGITUDE_DELTA
    }

    this.state = {
      mapRegion: region,
      selectedStation: selectedStation,
      hackLocationButton: 1
    }
  }

  onMarkerPress(e) {
    const destination = e.nativeEvent.coordinate
    this.directionsToolbar.show(destination)
  }

  onPress(e) {
    if (e.nativeEvent.action !== 'marker-press') { //user pressed anywhere else in the map,
      this.directionsToolbar.hide()
    }
  }

  componentWillMount() {
     //Hack to ensure the showsMyLocationButton is shown initially. Idea is to force a map repaint.
    setTimeout(()=> this.setState({
      hackLocationButton: 0
    }),500);
  }
  render() {
    const { stations } = this.props
    return (
      <View style={{flex: 1, marginTop: this.state.hackLocationButton}}>
        <MapView
          style={styles.map}
          region={this.state.mapRegion}
          showsUserLocation={true}
          toolbarEnabled = {false}
          zoomEnabled={true}
          onPress={this.onPress.bind(this)}
          >
          {stations.map(station => (
            <MapView.Marker
              coordinate={{
                latitude: station.latitude,
                longitude: station.longitude,
              }}
              image = {station.free_bikes == 0 ? require('../../assets/img/bikes_empty.png') : require('../../assets/img/citybike_logo.png')}
              title={`${station.extra.internal_id} ${station.name}`}
              description={`${station.free_bikes} bikes | ${station.empty_slots} empty slots`}
              key={station.id}
              ref={station.id}
              onPress={this.onMarkerPress.bind(this)}
              >
            </MapView.Marker>
          ))}
        </MapView>
        <DirectionsToolbar
          ref = { (ref) => this.directionsToolbar = ref }
        />
      </View>
    );
  }
}

export default Plan;
