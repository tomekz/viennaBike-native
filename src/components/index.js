import StationsList from './StationsList';
import StationCard from './StationCard';
import LoadingScreen from './LoadingScreen';
import Plan from './Plan';
import Header from './Header';

export {
  StationsList,
  StationCard,
  LoadingScreen,
  Plan,
  Header
};
