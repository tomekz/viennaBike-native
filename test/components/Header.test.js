import React from 'react';
import { Header } from '../../src/components'

import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(<Header/>).toJSON();
  expect(tree).toMatchSnapshot()
})

