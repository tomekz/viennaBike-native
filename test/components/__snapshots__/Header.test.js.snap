// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#76C2AF",
      "flexDirection": "row",
      "paddingBottom": 5,
      "paddingTop": 30,
    }
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "marginLeft": 15,
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    />
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "white",
        "flex": 5,
        "fontFamily": "Montserrat-Bold",
        "fontSize": 18,
        "margin": 1,
      }
    }
  >
    CityBike Vienna
  </Text>
</View>
`;
